### ISSUE TITLE
noop puppet runs with no changes reporting as 'noop' instead of 'unchanged' for puppetdb > 4.1.0

### ISSUE THREAD

**Issue by cintiadr** [voxpupuli/pypuppetdb#103]
Scenario: puppet is running as --noop every 30 minutes, and so far they were being reported in puppetboard as 'unchanged'; which is exactly what I expected, as no 'noops' events existed. It was pypuppetdb 0.3.1, puppetdb 4.0.x. 

Something example code to reproduce it:
```
from pypuppetdb import connect

puppetdb = connect(host='puppetdb', port=8080, timeout=20,)
nodes = puppetdb.nodes(with_status=True)
for node in nodes:
  print(node.name + ' - ' + puppet.status)
```

Looking at https://github.com/voxpupuli/pypuppetdb/blob/master/pypuppetdb/api.py#L413, it appears that I was getting in the 'else' clause, because 'latest_report_noop' was only included in puppetdb [4.1.0](https://docs.puppet.com/puppetdb/latest/release_notes.html#section-6):

```
                    if node['latest_report_noop']:
                        node['status'] = 'noop'
                    else:
                        node['status'] = node['latest_report_status']
```

I upgraded puppetdb from 4.0.x to 4.2.x. And suddenly now those same nodes are reporting as 'noop' regardless if there are changes or not to be applied. I suppose you wouldn't expect such change of behaviour, and in all fairness, puppetboard does treat 'noop' as 'pending changes'. 

Shouldn't it be using `latest_report_noop_pending` instead of `latest_report_noop`?

In case it helps, an example output in puppetdb 4.2.x, after noop without changes. 
```
  {
    "deactivated": null,
    "latest_report_hash": "02f848510cab4b519a58d3698ef9b07459914f44",
    "facts_environment": "production",
    "cached_catalog_status": "not_used",
    "report_environment": "production",
    "latest_report_corrective_change": null,
    "catalog_environment": "production",
    "facts_timestamp": "2016-08-30T13:18:28.466Z",
    "latest_report_noop": true,
    "expired": null,
    "latest_report_noop_pending": false,
    "report_timestamp": "2016-08-30T13:18:27.019Z",
    "certname": "9745e1389f91",
    "catalog_timestamp": "2016-08-30T13:18:28.856Z",
    "latest_report_status": "unchanged"
  }
```

**Comment by corey-hammerton** [comment]
And now I'm convinced to add `noop_pending to the various types. Please see #101

**Comment by cintiadr** [comment]
If I understood it correctly, the logic for puppetdb < 4.1.0 is:

On noop runs:
  - failed: something quite unhappy happened here, like a wrong resource status test
  - noop: run with pending items 
  - unchanged: run without pending items 


On regular runs:
  - failed: run with a failure 
  - changed: run with only successes 
  - unchanged: run without changes attempted


So the ambiguous state was 'unchanged', not really 'noop'. What happened is that 'unchanged' is not ambiguous any more, but now 'noop' is (it can mean either no changes pending or changes pending for noop runs). As the applications using this library cannot tell which version of puppetdb is running, it's hard to tell if 'noop' status can mean only pending changes or if it's the ambiguous one which means two things. 


I think it's important to maintain the consistence for consumers of the API, regardless of the version of puppetdb we are using. My opinion is that we should keep 'noop' status for only pending changes, and keep 'unchanged' ambiguous as the first fix. I'm happy to do this change. 

About improvements, I suppose renaming 'noop' to 'pending_changes' could be a good step, but I'd assume this is a breaking change. Also, it's possible to split 'unchanged' between 'unchanged' and 'unchanged_noop' or something.

**Comment by corey-hammerton** [comment]
What we can do for the Report and Node type is this:

- If `noop` == `false` use that status field value (existing logic)
- If `noop` == `true` and the status field == `unchanged` set the object's status field to `unchanged`
- If `noop` == `true` and the status field != `unchanged` set the object's status field to `noop`

This would be released with 0.3.2 release and would still be backward compatible. Mind you the new `noop_pending` fields in PuppetDB >= 4.2.0 appear to cover this logic but relying strictly on these fields may break compatibility in PuppetDB < 4.2.0. We will maintain compatibility between minor versions and implement the breaking changes between major versions only (if released).

Does this fulfill your requirements?

**Comment by cintiadr** [comment]
It absolutely does. 

Created the PR, but I couldn't find the tests for the node with_status tests. 

What will happen is that Failures in noop runs (as, for example, just got one when doing 'zypper list-updates'), it will show up as 'noop' instead of the previous 'Failed'. I have no problem with that, as they both need attention.

**Comment by cintiadr** [comment]
Actually no, scratch that. 

If you run a noop, regardless if it will trigger changes or not, puppetdb returns on /nodes 'latest_report_noop = true' and 'latest_report_status = unchanged'. 'latest_report_noop_pending' is the only thing that appears to change. 

On the reports/ endpoint, same deal. 'noop_pending' appear to be the only relevant difference. They were both included in 4.1.0, no? 

If so, the code was already relying on the other path, which will go by events.

**Comment by cintiadr** [comment]
Thing is I couldn't really find easily any vagrant box or docker with puppetdb 3.x and 4.0.x, but looking at the code it feels to me it would work just like before, no? Because noop was only added on 4 series. 

(I opened the pull request so it makes it simpler to comment)
